op-service/eth/transactions.go

```diff
@@ -42,3 +50,48 @@ func TransactionsToHashes(elems []*types.Transaction) []common.Hash {
 	}
 	return out
 }
+
+// CheckRecentTxs checks the depth recent blocks for transactions from the account with address addr
+// and returns the most recent block and true, if any was found, or the oldest block checked and false, if not.
+func CheckRecentTxs(
+	ctx context.Context,
+	lgr log.Logger,
+	l1 L1Client,
+	depth int,
+	addr common.Address,
+) (recentBlock uint64, found bool, err error) {
+	blockHeader, err := l1.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return 0, false, fmt.Errorf("failed to retrieve current block header: %w", err)
+	}
+
+	currentBlock := blockHeader.Number
+	currentNonce, err := l1.NonceAt(ctx, addr, currentBlock)
+	if err != nil {
+		return 0, false, fmt.Errorf("failed to retrieve current nonce: %w", err)
+	}
+
+	oldestBlock := new(big.Int)
+	oldestBlock.Sub(currentBlock, big.NewInt(int64(depth)))
+	previousNonce, err := l1.NonceAt(ctx, addr, oldestBlock)
+	if err != nil {
+		return 0, false, fmt.Errorf("failed to retrieve previous nonce: %w", err)
+	}
+
+	if currentNonce == previousNonce {
+		lgr.Info("No recent batcher txs detected. No need to wait for rollup sync")
+		return oldestBlock.Uint64(), false, nil
+	}
+
+	lgr.Info("Recent tx detected. Searching for block with most recent tx")
+	// Decrease block num until we find the block before the most recent batcher tx was sent
+	targetNonce := currentNonce - 1
+	for currentNonce > targetNonce {
```

Comment:
I just realize, there's a very unlikely yet possible edge-case that a reorg happens while querying `previousNonce, err := l1.NonceAt(ctx, addr, oldestBlock)` where one tx would be added, so that the `currentNonce == previousNonce` check would fail, then another reorg removes that tx again, and then this loop would potentially run for a very long time. It's a very constructed case, but I think just to be on the safe side we should also test the current block number
```suggestion
	for currentNonce > targetNonce && currentBlock.Cmp(oldestBlock) != -1 {
```
We could add a test case for this that would currently fail where the first nonce query for block N returns x, then querying the nonce for N-depth returns x+1, then all subsequent queries return x again.
Maybe there are other edge cases too that reorgs might cause that I don't currently see.

Reply:
I'm trying to understand the reorg situation you described. This is the nonce setting at each block before the reorg:
```
Blocks       495 496 497 498 499 500
Nonce          5   5   5   5   6   6
call NonceAt   x   -   -   -   -   x
```
Reorg occurs, removing tx that caused nonce to increment to 6. The nonce now looks like this:
```
Blocks       495 496 497 498 499 500
Nonce          5   5   5   5   5   5
call NonceAt   x   -   -   -   x   -
```
With the current implementation, we would start walking back from 500 to 499, then detect that the nonce at 499 is less than the nonce we previously read at 500, so `CheckForRecentTxs` would assume the nonce change occurred at block 500 and return that value. Did I describe the situation you had in mind correctly or am I missing something?

Reply:
I'm talking about a reorg `depth` deep, so
* 500 -> 5
* now reorg a tx into block 495
* 495 -> 6
* now reorg this tx out again
* 499 -> 5
* 498 -> 5
* ... potentially going on with 5s for a long time.

Reply:
Updated in https://github.com/ethereum-optimism/optimism/pull/10193/commits/4abebbcc3e19a40c864a506fcc159e197e1d06ee